contributors/design-proposals/public-identities-for-pods.md

```diff
@@ -0,0 +1,132 @@
+* [Implementation](#implementation)
+* [Alternatives](#alternatives)
+* [History](#history)
+
+# Goals
+
+* The main goal of Public Identities for Pods is to expose each Pod as an external IP address.
+  This approach allows sophisticated applications to communicate with the outside world using custom protocols.
+  A examples which require this approach are:
+  * Game Servers (which often use RTP, UDP)
+  * Data Grids (with L7 routing, such as [Infinispan](http://infinispan.org/docs/dev/user_guide/user_guide.html#using_hot_rod_server))
+  * Media Servers
+
+* The secondary goal is to create an API similar to `LoadBalancer` to manage external IP addresses.
+
+* The implementation is targeted (but not limited to) StatefulSets. Stable hostnames allow to identify the same
+  Stateful Pod after crash or restart (it should have a stable hostname). However it also makes sense to expose
+  ordinary deployments this way.
+
+# Existing solution
+
+* The above goal is partially possible to achieve with Loadbalanced Services, NodePort Services or
+  Ingress with session stickiness (if HTTP is the main protocol used for communication) but very often
+  this is not good enough and application clients assume there is only one application instance
+  behind an IP address (in other words there 1 to 1 mapping between IP addresses and Application
+  instances). Therefore a specialized implementation exposing an IP address per Pod is required.
+
+# API
+
+A new type of Service needs to be created in order to indicate that an external IP address will be exposed per Pod:
+
+```go
+const (
+    //Existing Service types
+	ServiceTypeClusterIP ServiceType = "ClusterIP"
+	ServiceTypeNodePort ServiceType = "NodePort"
+	ServiceTypeLoadBalancer ServiceType = "LoadBalancer"
+	ServiceTypeExternalName ServiceType = "ExternalName"
+
+	//new Service type - expose IP per Pod
+	ServiceTypeIPPerPod ServiceType = "IPPerPod"
+)
+```
+
+A new interface in `cloud.go` will be created to allow creating an external IP addresses.
+
+```go
+// ExternalIP is an abstract, pluggable interface for external ips.
+type ExternalIP interface {
+	// GetExternalIP returns whether the specified external IP exists, and
+	// if so, what its status is.
+	// Implementations must treat the *v1.Service parameter as read-only and not modify it.
+	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+	GetExternalIP(clusterName string, service *v1.Service) (status *v1.ExternalIPStatus, exists bool, err error)
+	// EnsureExternalIP creates a new external IP 'name', or updates the existing one.
+	// Returns the status of the external IP.
+	// Implementations must treat the *v1.Service and *v1.Node
+	// parameters as read-only and not modify them.
+	// Parameter 'clusterName' is the name of the cluster as presented to kube-controller-manager
+	EnsureExternalIP(clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.ExternalIPStatus, error)
```

Comment:
I think I made a mistake here. I wanted to pass a clusterName, a Service (which governs a clustered app deployed in Kubernetes, such as Data Grid) and a list of Pod (not Nodes, I'm sorry). I'm pretty sure the cloud vendor could easily extract a list of Pods from governing Service, so let me remove the last argument.

The reason for returning a status here is that some cloud vendors might not implement this api and we will return an error here. Then we could fall back to abusing Load Balancers for the instance.